File: matlab/test_shock_absorber_tt.m
% Shock absorber tests with TT decomposition
% The following parameters can be passed in the form 
% 'param_name1', param_value1, 'param_name2', param_value2 and so on:
%   x: a matrix of covariates (can be empty, in this case drawn from normal
%      distribution)
%   log2N: log2(number of samples in the chain)
%   D: number of covariates
%   delta: TT approximation/stopping threshold
%   n: number of grid points in each variable
%   runs: number of runs of the test (must be >1 for error estimation)
% The test script does not require output parameters. Instead, it copies
% all variables to the main Matlab workspace. Among those are:
%   N_cross:  number of pi evaluations in different runs of TT Cross
%   ttimes_cross: CPU times of TT cross runs
%   ttimes_invcdf: CPU times of IRT runs
%   num_of_rejects: numbers of rejections in MH method
%   Q_mh: quantiles computed by MH method
%   Q_iw: quantiles computed by IW method with QMC
%   tauint_tt: integrated autocorrelation times in each run
%   err_TT: estimate of the TT approximation error (empirical std)
function test_shock_absorber_tt(varargin)
% Check prerequisites
check_tt;
check_qmc;
check_mcmc;

params = parse_shock_inputs(varargin{:});
% extra parameters for TT
if (~isfield(params, 'n'))
    params.n = input('Uniform grid size for the posterior n = ? (default 16): ');
    if (isempty(params.n))
        params.n = 16;
    end
end
if (~isfield(params, 'delta'))
    params.delta = input('TT approx. threshold for the posterior delta = ? (default 0.05): ');
    if (isempty(params.delta))
        params.delta = 0.05;
    end
end

% Data table from the paper
y = [6700 6950 7820 8790 9120 9660 9820 11310 11690 11850 11880 12140 12200 ...
     12870 13150 13330 13470 14040 14300 17520 17540 17890 18420 18960  ...
     18980 19410 20100 20100 20150 20320 20900 22700 23490 26510 27410  ...
     27490 27890 28100];
censind = [0 1 1 1 0 1 1 1 1 1 1 1 0 1 0 1 1 1 0 0 1 1 1 1 1 1 0 1 1 1 0 0 1 0 1 0 1 1];

fprintf('Shock absorber TT test:\n');
% check for covariates data
if (isempty(params.x))
    % Simulate artificial covariates
    params.x = randn(params.D, numel(y))*1/params.D;
    fprintf('\tcovariates: randomly synthesized, mean(x)=%g\n', mean(params.x(:)));
else
    fprintf('\tcovariates: given, mean(x)=%g\n', mean(params.x(:)));
end
fprintf('\tNumber of samples: N=2^%d\n', params.log2N);
fprintf('\tNumber of covariates: D=%d\n', params.D);
fprintf('\tTT approx. tolerance: delta=%g\n', params.delta);
fprintf('\tNumber of grid points: n=%d\n', params.n);
fprintf('\tNumber of runs: runs=%d\n', params.runs);

% scales for beta
beta_mean = zeros(1,params.D+1);
beta_mean(1) = log(30796);
beta_var = ones(1,params.D+1);
beta_var(1) = 0.1563;

% Ranges of domain
a = beta_mean - 3*sqrt(beta_var);
b = 2*beta_mean - a;
a = [a 0]';
b = [b 13]';

% Vectors of grid points
theta = cell(params.D+2,1);
for i=1:params.D+2
    h = (b(i)-a(i))/params.n;
    theta{i} = (a(i)+h:h:b(i))';
    theta{i} = tt_tensor(theta{i});
end
% Replicate those in all dimensions as TT tensors
% These must be without boundary points
theta_tt = tt_meshgrid_vert(theta);
% Add boundary points for the Rosenblatt transform
for i=1:params.D+2
    theta{i} = full(theta{i});
    theta{i} = [a(i); theta{i}];
end

% Exact posterior function
pifun = @(theta)exp(shock_log_weibull(theta, params.x, y, censind) + shock_log_prior(theta, beta_mean, beta_var));

n_err_TT = round(params.runs/4); % number of runs used to estimate error in pi
Pi = cell(1,n_err_TT); % storage for pi approximations from several runs

% Allocate storage
N_cross = zeros(params.runs,2);  % number of pi evaluations in TT Cross
ttimes_cross = zeros(params.runs, 1); % CPU time of TT cross
ttimes_invcdf = zeros(params.runs, 1); % times of IRT
num_of_rejects = zeros(params.runs, 1); % number of rejections in MH
Q_mh = zeros(params.runs, 2); % quantiles computed by MH method
Q_iw = zeros(params.runs, 2); % quantiles computed by IS method with QMC
tauint_tt = zeros(params.runs, 1); % IACT (after MH)


for irun=1:params.runs
    tic;
    [pi,~,~,~,N_cross(irun,:)] = amen_cross_s(theta_tt, pifun, 0, 'kickrank', 2, 'y0', 8, 'tol_exit', params.delta);
    ttimes_cross(irun) = toc;
    pi = round(pi, 1e-6); % truncate some ranks we can
    if (irun<=n_err_TT)
        Pi{irun} = pi;
    end

    [Z_mh,~,num_of_rejects(irun),ttimes_invcdf(irun)] = tt_irt_debias(2^params.log2N, pifun, pi, theta, 'mcmc');
    [Q_mh(irun,:)] = shock_quantiles(Z_mh, 0*ones(params.D,1));

    % Importance weighting with QMC
    Z_iw = qmcnodes(params.D+2, params.log2N)';
    [Z_iw,pi_app] = tt_irt(theta,pi,Z_iw);
    pi_ex = pifun(Z_iw);
    Q_iw(irun,:) = shock_quantiles(Z_iw, 0*ones(params.D,1), pi_ex./pi_app);

    % Estimate autocorr time of Q1
    [~,~,~,tauint_tt(irun),~,~] = UWerr(Z_mh,1.5,length(Z_mh),0);
    tauint_tt(irun) = tauint_tt(irun)*2;    
end
N_cross = N_cross(:,2);

% Estimate the actual TT approximation error
n_err_TT = min(n_err_TT,params.runs);
err_TT = nan;
if (n_err_TT>1)
    pi = amen_sum(Pi, (1/n_err_TT)*ones(n_err_TT,1), 1e-6, 'fkick', true, 'y0', Pi{1}, 'kickrank', 16);
    err_TT = norm(pi)*ones(n_err_TT,1);
    for irun=1:n_err_TT
        err_TT(irun) = norm(Pi{irun} - pi)/err_TT(irun);
    end
    err_TT = sqrt(sum(err_TT.^2)/(n_err_TT-1)); % empirical standard deviation
end


% Print some statsy information
fprintf('TT Shock absorber completed. Some average values:\n');
fprintf('\tN_cross: %g\n', mean(N_cross));
fprintf('\tttimes_cross: %g\n', mean(ttimes_cross));
fprintf('\tttimes_invcdf: %g\n', mean(ttimes_invcdf));
fprintf('\tnum_of_rejects: %g (out of N=%d)\n', mean(num_of_rejects), 2^params.log2N);
fprintf('\tIACT: %g\n', mean(tauint_tt));
fprintf('\tTT error: %g\n', err_TT);
fprintf('\tQ_mh: [%g %g]+-[%g %g]\n', mean(Q_mh(:,1)), mean(Q_mh(:,2)), sqrt(sum((Q_mh(:,1)-mean(Q_mh(:,1))).^2)/(params.runs-1)),  sqrt(sum((Q_mh(:,2)-mean(Q_mh(:,2))).^2)/(params.runs-1)));
fprintf('\tQ_iw: [%g %g]+-[%g %g]\n', mean(Q_iw(:,1)), mean(Q_iw(:,2)), sqrt(sum((Q_iw(:,1)-mean(Q_iw(:,1))).^2)/(params.runs-1)),  sqrt(sum((Q_iw(:,2)-mean(Q_iw(:,2))).^2)/(params.runs-1)));

% Copy vars to main space
vars = whos;
for i=1:numel(vars)
    if (exist(vars(i).name, 'var'))
        assignin('base', vars(i).name, eval(vars(i).name));
    end
end
end

File: python/test_shock_absorber_tt.py
from __future__ import print_function, absolute_import, division
import numpy as np
import tt
import tt.cross
from tt.cross.rectcross import rect_cross
from tt_irt import tt_irt1
import time
import sys

try:
    import puwr
except:
    print("Couldn't import puwr, you may correct print statements there for Python3")

# log-prior distribution
def logf_prior(theta, beta_mean, beta_var):
    d = theta.shape[1]-2
    I = theta.shape[0]

    alpha = 6.8757
    beta = 2.2932
    theta2 = theta[:,d+1]
    theta2 = np.reshape(theta2, [I,1], order='F')

    beta_mean_I = np.repeat(np.reshape(beta_mean, [1,d+1]), I, axis=0)
    beta_mean_I = np.reshape(beta_mean_I, [I, d+1], order='F')
    beta_var_I = np.repeat(np.reshape(beta_var, [1,d+1]), I, axis=0)
    beta_var_I = np.reshape(beta_var_I, [I, d+1], order='F')
    theta2_d = np.repeat(theta2, d+1, axis=1)
    theta2_d = np.reshape(theta2_d, [I, d+1], order='F')
    theta_in = theta[:, range(0,d+1)]
    theta_in = np.reshape(theta_in, [I, d+1], order='F')

    F = -((theta_in - beta_mean_I)**2)*0.5*theta2_d/beta_var_I
    F = np.sum(F, axis=1)
    F = np.reshape(F, [I, 1], order='F')
    F = F + (alpha-0.5)*np.log(theta2) - beta*theta2
    return F


# log-Likelihood
def logL_weibull(theta, x, y, c):
    d = theta.shape[1]-2
    I = theta.shape[0]
    beta = theta[:, range(0,d+1)]
    beta = np.reshape(beta, [I, d+1], order='F')
    lam = np.reshape(theta[:, d+1], [I,1], order='F')
    F = np.zeros(I, dtype=np.float64)
    F = np.reshape(F, [I,1], order='F')
    m = np.size(y)
    x = np.reshape(x, [d,m], order='F')
    beta0 = np.reshape(beta[:,0], [I,1], order='F')
    for i in range(0,m):
        logeta = beta[:,range(1,d+1)].dot(x[:,i])
        logeta = np.reshape(logeta, [I,1], order='F')
        logeta = logeta + beta0
        eta = np.exp(logeta)
        yeta = y[i]/eta
        if c[i] == 1:  # censored
            f = -(yeta**lam)
        else:
            f = np.log(lam) - logeta + (lam-1.0)*(np.log(y[i]) - logeta) - (yeta**lam)
            f = f + np.log(30000.0) # to prevent underflow

        F = F + f
    return F


# Full index function for cross
def crossfun(ind, theta, x, y, censind, beta_mean, beta_var):
    d = ind.shape[1]
    I = ind.shape[0]

    # Cross indices are integers mathematically,
    # but in rect_cross they are sometimes stored as doubles.
    # Fix that
    ind = ind.astype(np.int32)

    thetai = np.zeros(I*d)
    thetai = np.reshape(thetai, [I,d], order='F')
    for i in range(0,d):
        thetai[:,i] = theta[i][ind[:,i]]
    thetai = np.reshape(thetai, [I,d], order='F')
    # Posterior density
    F = np.exp(logf_prior(thetai, beta_mean, beta_var) + logL_weibull(thetai, x, y, censind))
    F = np.reshape(F, [I,1], order='F')
    return F


###################### MAIN ###############

d = 6       # number of covariates
n = 17      # grid size
tol = 5e-2  # TT stopping threshold
log2N = 16  # log2(number of samples)
runs = 8    # number of runs

y = [6700, 6950, 7820, 8790, 9120, 9660, 9820, 11310, 11690, 11850, 11880, 12140, 12200,
     12870, 13150, 13330, 13470, 14040, 14300, 17520, 17540, 17890, 18420, 18960,
     18980, 19410, 20100, 20100, 20150, 20320, 20900, 22700, 23490, 26510, 27410,
     27490, 27890, 28100]
censind = [0, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0, 1, 1]

# Estimate the ranges of beta
beta_mean = np.zeros(d+1)
beta_mean[0] = np.log(30796.0)
beta_var = np.ones(d+1)
beta_var[0] = 0.1563

a = beta_mean - 3.0*np.sqrt(beta_var)
b = 2.0*beta_mean - a
a = np.hstack([a, 0.0])
b = np.hstack([b, 13.0])

# Vectors of grid nodes
theta = (d+2)*[None]
theta_f = (d+2)*[None]
for i in range(0,d+2):
    h = (b[i]-a[i])/(n-1)
    theta[i] = a[i]+np.arange(0,n)*h
    theta_f[i] = np.reshape(theta[i], [n, 1], order='F')

theta_f = np.hstack(theta_f)
theta_f = np.reshape(theta_f, [(d+2)*n, 1], order='F')

# Simulate artificial covariates
x = np.random.randn(d*np.size(y))/d;
x = np.reshape(x, [d, np.size(y)], order='F')

# short index function for cross
myfun = lambda ind: crossfun(ind, theta, x, y, censind, beta_mean, beta_var)

print('Running '+repr(runs)+' tests for '+repr(d)+' covariates, producing '+repr(2**log2N)+' samples')
print('')

Q_py = runs*[None]
tau = runs*[None]
for irun in range(0,runs):
    # Run cross
    f0 = tt.rand(n, d+2, 8)
    f = rect_cross.cross(myfun, f0, nswp=50, kickrank=2, rf=2, eps=tol)
    print(f)


    # Seed points
    M = 2**log2N
    q = np.random.random([M, d+2])
    q = np.reshape(q, [M, d+2], order = 'F')


    ttt = time.time()
    # Sample
    Z, lPz = tt_irt1(q, f, theta_f)
    ttimes_invcdf = time.time() - ttt
    print('Sampling time = '+repr(ttimes_invcdf))

    Z = np.reshape(Z, [M,d+2], order='F') # Proposed samples
    lPz = np.reshape(lPz, [M,1], order='F') # Approximate log(density)

    # Exact log(density)
    lPex = logL_weibull(Z, x, y, censind) + logf_prior(Z, beta_mean, beta_var)

    # MCMC rejections
    num_of_rejects = 0
    for i in range(0,M-1):
        alpha = np.exp(lPex[i+1] - lPex[i] + lPz[i] - lPz[i+1])
        if alpha<np.random.random(1): # reject
            Z[i+1,:] = Z[i,:]
            lPex[i+1] = lPex[i]
            lPz[i+1] = lPz[i]
            num_of_rejects = num_of_rejects + 1

    print('rejection rate = ' + repr((num_of_rejects*100.0)/M))

    # QoI = Quantile function
    q = 0.05
    theta1 = np.exp(Z[:,0])
    theta2 = Z[:,d+1]
    q_post = theta1*((-np.log(q))**(1.0/theta2))
    # IACT -- only works in python 2.7 due to bad print
    try:
        dumpmean, delta, tint, d_tint = puwr.tauint(np.reshape(Z.T, [d+2, 1, M]), 0)
        tau[irun] = tint*2.0
    except:
        pass


    print('tau = ' + repr(tau[irun]))
    q_post = np.mean(q_post)
    print('Mean quantile = ' + repr(q_post))
    Q_py[irun] = q_post


print('')
print('TT Shock absorber completed. Some average values:')
print('Q_py = '+repr(np.mean(Q_py))+' +- '+repr(np.sqrt(np.sum((Q_py - np.mean(Q_py))**2)/(runs-1))) )
try:
    print('tau = '+repr(np.mean(tau))+' +- '+repr(np.sqrt(np.sum((tau - np.mean(tau))**2)/(runs-1))) )
except:
    pass


